connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java

```diff
@@ -231,5 +223,45 @@ public synchronized Void get(long timeout, TimeUnit unit) throws InterruptedExce
         }
     }
 
+    private class KafkaOffsetReadFuture extends ConvertingFutureCallback<Void, Map<ByteBuffer, ByteBuffer>> implements OffsetReadFuture {
+        private final Collection<ByteBuffer> keys;
+        private final AtomicBoolean completed;
+
+        public KafkaOffsetReadFuture(Collection<ByteBuffer> keys) {
+            super(null);
+            this.keys = keys;
+            this.completed = new AtomicBoolean(false);
+        }
+
+        @Override
+        public Map<ByteBuffer, ByteBuffer> convert(Void result) {
+            Map<ByteBuffer, ByteBuffer> values = new HashMap<>();
+            for (ByteBuffer key : keys)
+                values.put(key, data.get(key));
+            return values;
+        }
 
+        @Override
+        public void prematurelyComplete() {
+            log.debug("Prematurely completing offset read future");
+            onCompletion(null, null);
```

Comment:
we should cache the fact that this future has been force stopped, and if so, any subsequent calls to get() should immediately return with either en empty map. right now, it looks like it will re-try to catch up with the remote log.

Reply:
I agree with the gist of this comment, but I think the right place for that caching behavior is in the offset reader class, not in the backing store (since each task is given its own offset reader, but the backing store is shared among all of them).